### PR TITLE
internal/state/rollback: get rollback metadata at pred height

### LINKED ISSUE #7304: rollback leads to an invalid state
Related issue: #7123

**Tendermint version**: master and `0.34.14`

**ABCI app** (name for built-in, URL for self-written if it's publicly available): https://github.com/anoma/anoma

**What happened**: I'm trying to use the new `tendermint rollback` version with the Anoma app. However, it seems that the command leads to an invalid state in Tendermint in the replay of the block:

```
Error: error during handshake: error on replay: block time 2021-11-22 14:22:01.19637851 +0000 UTC not greater than last block time 2021-11-22 14:22:01.19637851+0000 UTC
```

Note that we are also rolling back the apps state, however the error is coming Tendermint's side with Tendermint's node state.

The `LastBlockTime` in the `State` seems to be using the value as the current block (i.e. it's not being rolled back to the predecessor block).

**What you expected to happen**: The Tendermint state should roll back to a consistent predecessor block from which it's possible to restart the node and replay the last block.

**Have you tried the latest version**: yes

**How to reproduce it** (as minimally and precisely as possible): Run rollback and restart the node

### PR DESCRIPTION
**Pull request by tzemanovic**
Closes #7304

### CODE REVIEW

**Review by cmwaters — APPROVED** [review]
:+1:

File: internal/state/rollback.go [hunk 1]
```diff
@@ -37,7 +37,7 @@ func Rollback(bs BlockStore, ss Store) (int64, []byte, error) {
 
 	// state store height is equal to blockstore height. We're good to proceed with rolling back state
 	rollbackHeight := invalidState.LastBlockHeight
-	rollbackBlock := bs.LoadBlockMeta(rollbackHeight)
+	rollbackBlock := bs.LoadBlockMeta(rollbackHeight - 1)
```
Comment (tzemanovic) [inline internal/state/rollback.go:40]: ah, this should really be
```
rollbackHeight := invalidState.LastBlockHeight - 1
rollbackBlock := bs.LoadBlockMeta(rollbackHeight)
```

and we'll need to update the calls below that are using the `rollbackHeight`

Comment (tzemanovic) [reply internal/state/rollback.go:40]: Line 45 `previousValidatorSet, err := ss.LoadValidators(rollbackHeight - 1)` is using the pred block height. Should `previousParams, err := ss.LoadConsensusParams(rollbackHeight)` (line 50) be using the same or is that right?

**Comment by cmwaters** [comment]
Hi @tzemanovic, I've merged your changes into this PR: https://github.com/tendermint/tendermint/pull/7315 which includes an integration test running the rollback function on the e2e application. I'm going to close this PR. Thanks for the contribution! It would also be great if you could still test out the new changes.

**Comment by tzemanovic** [comment]
> Hi @tzemanovic, I've merged your changes into this PR: #7315 which includes an integration test running the rollback function on the e2e application. I'm going to close this PR. Thanks for the contribution! It would also be great if you could still test out the new changes.

Thank you too, I'll try it out!

**Pull request closed without merging**